File: romanToInteger/fastestSolution.cpp
// link: https://leetcode.com/problems/roman-to-integer/
class Solution {
    public int romanToInt(String s) {
        int sum = 0;
        
        int n = s.length();
        for (int i = n-1; i>=0; i--){
            char ch = s.charAt(i);
            switch (ch){
                case 'I': sum += sum>=5?-1: 1; break;
                case 'V': sum += 5;break;
                case 'X': sum += sum>=50?-10: 10;break;
                case 'L': sum += 50;break;
                case 'C': sum += sum>=500? -100: 100;break;
                case 'D': sum += 500;break;
                case 'M': sum += 1000;break;
                
            }
        }
        
        return sum;
    }
}

File: findLongestWord/fastestSolution.java
// link: https://leetcode.com/problems/longest-word-in-dictionary-through-deleting/
class Solution {
    public String findLongestWord(String s, List<String> d) {
        int len = s.length();
        
        int[][] indexes = new int[len + 1][26];
        
        Arrays.fill(indexes[len], Integer.MAX_VALUE);
        
        for (int i = len - 1; i >= 0; i--) {
            for (int j = 0; j < 26; j++) {
                indexes[i][j] = indexes[i + 1][j];
            }
            indexes[i][s.charAt(i) - 'a'] = i + 1;
        }
        
        String res = "";
        
        for (String word : d) {
            if (word.length() > s.length() || word.length() < res.length()) continue;
            int i = 0;
            int j = 0;
            
            while (j < word.length()) {
                char c = word.charAt(j);
                if (indexes[i][c - 'a'] == Integer.MAX_VALUE) break;
                i = indexes[i][c - 'a'];
                j++;
            }
            
            if (j == word.length()) {
                if (word.length() > res.length() || word.length() == res.length() && word.compareTo(res) < 0) {
                    res = word;
                }
            }
        }
        
        return res;
    }
}